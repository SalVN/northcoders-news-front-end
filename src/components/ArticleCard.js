import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

const ArticleCard = function (props) {
  return (
    <div className='box'>
      <article className='media'>
        <div className='media-left'>
          {props.number + 1}
        </div>
        <div className='media-left'>
          <button className='button' onClick={props.voteHandler.bind(this, 'up', props.id)}><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
          <div>{props.votes}</div>
          <button className='button' onClick={props.voteHandler.bind(this, 'down', props.id)}><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
        </div>
        <div className='media-content'>
          <div className='content'>
            <Link to={`/articles/${props.id}`}>
              <h3 className='title is-3'>{props.title}</h3>
            </Link>
            <p>
              <span>added by: {props.author}</span>
              <span>tags: {props.tags}</span>
              <span>comments: {props.comment_count}</span>
            </p>
          </div>
        </div>
      </article>
    </div>
  );
};

ArticleCard.propTypes = {
  title: PropTypes.string.isRequired,
  votes: PropTypes.number.isRequired,
  number: PropTypes.number.isRequired,
  author: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  tags: PropTypes.string.isRequired,
  comment_count: PropTypes.number.isRequired,
  voteHandler: PropTypes.func.isRequired
};


export default ArticleCard;
