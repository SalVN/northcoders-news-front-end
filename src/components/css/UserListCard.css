.card-box {
    margin: 10px;
}

.card-header-style {
    background-color: #c1002e;
    min-height: 50px;
    max-height: 100px;
    width: 100%;
}

.card-header-title {
    color: white !important;
}

.user-list-card-number {
    color: black;
    font-size: 16px;
}

.user-list-image {
    width: 24px;
    height: 24px;
    margin-right: 5px;
}

.user-list-username-text {
    color: #c1002e;
    font-weight: 600;
    font-size: 14px;
    margin-top: -10px;
    margin-bottom: 5px;
}

.user-list-card-ranking-number {
    color: black;
    font-weight: 800;
    font-size: 16px;
}

.mini-card {
    border-bottom: 1px solid lightgrey;
    margin-bottom: 5px;
}

.dropdown-button {
    border: 0;
    margin-right: 10px;
}

.cancel {
    color: #c1002e;
    font-weight: 500;
}