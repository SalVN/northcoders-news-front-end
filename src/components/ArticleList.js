import React from 'react';
import ArticleCard from './ArticleCard';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';

class ArticleList extends React.Component {
  componentDidMount() {
    if (this.props.users && this.props.users.length < 1) {
      this.props.fetchUsers();
    }
  }
  render() {
    const users = this.props.users;
    const voteHandler = this.props.voteArticle;
    return (
      <div id='ArticleList' className='content'>
        {this.props.articles.length > 0 &&
          this.props.articles.map((article, i) => {
            const username = article.created_by;
            let index;
            users &&
              users.forEach((user, i) => {
                if (user.username === username) index = i;
              });
            return (
              <ArticleCard
                key={article._id}
                title={article.title}
                number={i}
                votes={article.votes}
                author={article.created_by}
                id={article._id}
                tags={article.belongs_to}
                comment_count={article.comment_count}
                voteHandler={voteHandler}
                userData={users[index]}
              />
            );
          })
        }
      </div>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    fetchUsers: () => {
      dispatch(actions.fetchUsers());
    }
  };
}

function mapStateToProps(state) {
  return {
    loading: state.users.loading,
    users: state.users.users
  };
}

ArticleList.propTypes = {
  articles: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchUsers: PropTypes.func.isRequired,
  voteArticle: PropTypes.func.isRequired,
  users: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ArticleList);