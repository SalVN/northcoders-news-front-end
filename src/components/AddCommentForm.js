import React from 'react';
import PropTypes from 'prop-types';

import './css/AddCommentForm.css';

class AddCommentForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: ''
        };
        this.handleChange = this.handleChange.bind(this);
    }
    render() {
        return (
            <div>
                {this.props.showForm
                    ?
                    <div className='box box-style box-style-comment box-style-comment'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <form onSubmit={this.props.handleSubmit}>
                                        <label className='comment-form-heading' htmlFor="comment-input">Join the conversation:</label>
                                        <br />
                                        <div className='comment-form' onBlur={this.props.toggleForm.bind(this, this.state.text)}>
                                            <textarea className='add-comment-textarea' onChange={this.handleChange} id="comment-input" defaultValue={this.state.text} placeholder='Join the conversation' />
                                        </div>
                                        <div className='button-div'>
                                            <button className='button add-comment-submit' type='submit'>Submit</button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </article>
                    </div>
                    : <div className='box box-style box-style-comment box-style-comment-small'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <div onClick={this.props.toggleForm.bind(this, null)} className='comment-form-small'>
                                        <p>Join the conversation</p>
                                    </div>
                                </div>
                            </div>
                        </article>
                    </div>
                }
            </div>
        );
    }
    handleChange(e) {
        this.setState({
            text: e.target.value
        });
    }
}

AddCommentForm.propTypes = {
    id: PropTypes.string.isRequired,
    addComment: PropTypes.func.isRequired,
    toggleForm: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    showForm: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default AddCommentForm;