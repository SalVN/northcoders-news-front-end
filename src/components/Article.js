import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

import './css/Article.css';

const Article = function (props) {
    return (
        <div className='box'>
            <div className='columns'>
                <div className='column is-1'>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
                    <p className='comment-card-votes'><strong>{props.article.votes}</strong></p>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
                </div>
                <div className='column'>
                    <div className='title is-3 article-title'>{props.article.title}</div>
                    <p>
                        <span>
                            <span className='article-subheading-key'>added by:</span>
                            <Link to='#'>
                                <span>
                                    {props.user &&
                                        <img className='article-card-avatar' src={props.user.avatar_url} alt='user avatar' />
                                    }
                                </span>
                                <span className='article-subheading-value'>
                                    <strong>{props.article.created_by}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>tags:</span>
                            <Link to={`/topics/${props.article.belongs_to}/articles`}>
                                <span className='article-card-subheading-value'>
                                    <strong>{props.article.belongs_to}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>comments:</span>
                            {/*<Link to='#'>*/}
                            <span className='article-card-subheading-value-notlink'>
                                <strong>{props.article.commentCount}</strong>
                            </span>
                            {/*</Link>*/}
                        </span>
                    </p>
                    <div>{props.article.body}</div>
                </div>
            </div>
        </div>
    );
};

Article.propTypes = {
    article: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired
};

export default Article;