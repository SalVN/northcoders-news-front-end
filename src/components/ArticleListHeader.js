import React from 'react';
import PropTypes from 'prop-types';

import './css/MainArticleList.css';

const ArticleListHeader = function (props) {
    return (
        <div className='content'>
            <div className='columns'>
                <div className='column is-3'>
                    <h2 className='page-title'><strong>{props.title}</strong></h2>
                </div>
                <div className='column is-1 is-offset-5 label'>
                    <label htmlFor="Sort by">Sort by:</label>
                </div>
                <div className='column is-3'>
                    <div className='select is-primary select-menu'>
                        <select name="Sort by" onChange={props.handleClickSelect}>
                            <option value="votes">Popularity</option>
                            <option value="comments">Number of Comments</option>
                            <option value="random">Random</option>
                        </select>
                    </div>
                </div>
            </div>
        </div>
    );
};

ArticleListHeader.propTypes = {
    title: PropTypes.string.isRequired,
    handleClickSelect: PropTypes.func.isRequired
};

export default ArticleListHeader;