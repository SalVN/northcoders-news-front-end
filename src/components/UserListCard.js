import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';
import { sortUsers } from '../utilities/sortUsers';

import UserRankingCard from './UserRankingCard';
import './css/UserListCard.css';

class UserListCard extends Component {
    constructor(props) {
        super(props);
        this.state = {
            showDropdown: false,
            sortBy: 'Sort by ranking'
        };
        this.toggleDropdown = this.toggleDropdown.bind(this);
        this.sortUserList = this.sortUserList.bind(this);
    }
    componentDidMount() {
        if (this.props.users && this.props.users.length < 1) {
            this.props.fetchUsers();
        }
    }
    render() {
        if (this.props.users && this.props.users.length > 0) {
            sortUsers(this.props.users, this.state.sortBy);
        }
        let activeDropdown = this.state.showDropdown ? 'is-active' : '';
        let rankingActive = this.state.sortBy === 'Sort by ranking' ? 'is-active' : '';
        let commentsActive = this.state.sortBy === 'Sort by comments' ? 'is-active' : '';
        let commentVotesActive = this.state.sortBy === 'Sort by comment votes' ? 'is-active' : '';
        let articleVotesActive = this.state.sortBy === 'Sort by article votes' ? 'is-active' : '';
        return (
            <div className='card-box'>
                <div className='container-fluid'>
                    <div className='card'>
                        <header className='card-header'>
                            <div className='card-header-style'>
                                <div className='card-header-title'>
                                    <div className={`dropdown ${activeDropdown}`}>
                                        <div className="dropdown-trigger">
                                            <button onClick={this.toggleDropdown} className="button is-small is-primary is-inverted is-outlined dropdown-button" aria-haspopup="true" aria-controls="dropdown-menu">
                                                <span className="icon">
                                                    <i className="fa fa-caret-down" aria-hidden="true"></i>
                                                </span>
                                            </button>
                                        </div>
                                        <div className="dropdown-menu" id="dropdown-menu" role="menu">
                                            <div className="dropdown-content">
                                                <a onClick={this.sortUserList} className={`dropdown-item ${rankingActive}`}>
                                                    Sort by ranking
                                                </a>
                                                <a onClick={this.sortUserList} className={`dropdown-item ${commentsActive}`}>
                                                    Sort by comments
                                                </a>
                                                <a onClick={this.sortUserList} className={`dropdown-item ${commentVotesActive}`}>
                                                    Sort by comment votes
                                                </a>
                                                <a onClick={this.sortUserList} className={`dropdown-item ${articleVotesActive}`}>
                                                    Sort by article votes
                                                </a>
                                                <a onClick={this.toggleDropdown} className="dropdown-item cancel">
                                                    Close
                                                </a>
                                            </div>
                                        </div>
                                    </div>
                                    <span>User Ranking</span>
                                </div>
                            </div>
                        </header>
                        <div className='card-content'>
                            {
                                this.props.usersLoading &&
                                <div className='article-list-loading-icon'>
                                    <span>
                                        <i className='fa fa-refresh fa-spin' />
                                    </span>
                                </div>
                            }
                            {
                                this.props.users &&
                                this.props.users.map((user, i) => {
                                    return (
                                        <UserRankingCard key={user.username} user={user} criteria={this.state.sortBy} index={i} />
                                    );
                                })
                            }
                        </div>
                    </div>
                </div>
            </div>
        );
    }
    toggleDropdown() {
        this.setState({
            showDropdown: !this.state.showDropdown
        });
    }
    sortUserList(e) {
        e.preventDefault();
        this.setState({
            showDropdown: false,
            sortBy: e.target.innerText
        });
    }
}

function mapDispatchToProps(dispatch) {
    return {
        fetchUsers: () => {
            dispatch(actions.fetchUsers());
        }
    };
}

function mapStateToProps(state) {
    return {
        users: state.users.users,
        usersLoading: state.users.loading
    };
}

UserListCard.propTypes = {
    users: PropTypes.array.isRequired,
    fetchUsers: PropTypes.func.isRequired,
    usersLoading: PropTypes.bool.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(UserListCard);