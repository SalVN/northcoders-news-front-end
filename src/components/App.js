import React, {Component} from 'react';
import PropTypes from 'prop-types';

import TopicNavBar from './TopicNavBar';
import MainBanner from './MainBanner';
import UserCard from './UserCard';
import UserListCard from './UserListCard';

import './css/App.css';

export class App extends Component {
  render() {
    let pageStyling = window.innerWidth < 800 ? 'main-page-mobile' : 'main-page';
    return (
      <div className='page'>
        <TopicNavBar />
        <MainBanner />
        <div className={pageStyling}>
          <div className={'columns is-gapless'}>
            <div className='column is-three-quarters'>
              {this.props.children}
            </div>
            <div className='column is-one-quarter'>
              <div className='is-hidden-mobile'>
                <UserCard />
              </div>
              <UserListCard />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

App.propTypes = {
  children: PropTypes.element.isRequired
};

export default App;
