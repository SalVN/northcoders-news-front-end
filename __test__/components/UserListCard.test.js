import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';

const mockStore = configureStore();
const initialState = {};

import { UserListCard } from '../../src/components/UserListCard';

describe('UserListCard', () => {
    const users = [{
        _id: '59b1b18b327cce1fb043bdb1',
        username: 'user1',
        name: 'One',
        comment_count: 10,
        articles_vote_count: 30,
        comments_vote_count: 20,
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }, {
        _id: '59b1b18b327cce1fb043bdb2',
        username: 'user2',
        name: 'Two',
        comment_count: 5,
        articles_vote_count: 41,
        comments_vote_count: 25,
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }, {
        _id: '59b1b18b327cce1fb043bdb3',
        username: 'user3',
        name: 'Three',
        comment_count: 15,
        articles_vote_count: 39,
        comments_vote_count: 30,
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }, {
        _id: '59b1b18b327cce1fb043bdb2',
        username: 'user4',
        name: 'Four',
        comment_count: 20,
        articles_vote_count: 60,
        comments_vote_count: 15,
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }];

    it('is a function', () => {
        expect(typeof UserListCard).toEqual('function');
    });

    it('renders', () => {
        const store = mockStore(initialState);
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
            store={store}
        />);
        expect(wrapper.children().length).toEqual(1);
    });

    it('renders correctly', () => {
        const store = mockStore(initialState);
        const tree = renderer.create(
            <Provider store={store}>
                <MemoryRouter>
                    <UserListCard
                        users={users}
                        fetchUsers={x => x}
                        usersLoading={false}
                        store={store} />
                </MemoryRouter>
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('has an initial state', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.state('showDropdown')).toBe(false);
        expect(wrapper.state('sortBy')).toBe('Sort by ranking');
    });

    it('fetches the users on componentDidMount if users is an empty array', () => {
        const spy = sinon.stub();
        mount(
            <MemoryRouter>
                <UserListCard
                    users={users}
                    fetchUsers={spy}
                    usersLoading={false}
                />
            </MemoryRouter>);
        expect(spy.callCount).toBe(0);
        mount(
            <MemoryRouter>
                <UserListCard
                    users={[]}
                    fetchUsers={spy}
                    usersLoading={false}
                />
            </MemoryRouter>);
        expect(spy.callCount).toBe(1);
    });

    it('sorts the users according to the dropdown if users is not an empty array', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.find('UserRankingCard').nodes[0].key).toBe('user4');
        expect(wrapper.find('UserRankingCard').nodes[1].key).toBe('user3');
        expect(wrapper.find('UserRankingCard').nodes[2].key).toBe('user2');
        expect(wrapper.find('UserRankingCard').nodes[3].key).toBe('user1');

        wrapper.setState({ sortBy: 'Sort by comments' });
        expect(wrapper.find('UserRankingCard').nodes[0].key).toBe('user4');
        expect(wrapper.find('UserRankingCard').nodes[1].key).toBe('user3');
        expect(wrapper.find('UserRankingCard').nodes[2].key).toBe('user1');
        expect(wrapper.find('UserRankingCard').nodes[3].key).toBe('user2');

        wrapper.setState({ sortBy: 'Sort by comment votes' });
        expect(wrapper.find('UserRankingCard').nodes[0].key).toBe('user3');
        expect(wrapper.find('UserRankingCard').nodes[1].key).toBe('user2');
        expect(wrapper.find('UserRankingCard').nodes[2].key).toBe('user1');
        expect(wrapper.find('UserRankingCard').nodes[3].key).toBe('user4');

        wrapper.setState({ sortBy: 'Sort by article votes' });
        expect(wrapper.find('UserRankingCard').nodes[0].key).toBe('user4');
        expect(wrapper.find('UserRankingCard').nodes[1].key).toBe('user2');
        expect(wrapper.find('UserRankingCard').nodes[2].key).toBe('user3');
        expect(wrapper.find('UserRankingCard').nodes[3].key).toBe('user1');

        wrapper.setState({ sortBy: 'Sort by ranking' });
        expect(wrapper.find('UserRankingCard').nodes[0].key).toBe('user4');
        expect(wrapper.find('UserRankingCard').nodes[1].key).toBe('user3');
        expect(wrapper.find('UserRankingCard').nodes[2].key).toBe('user2');
        expect(wrapper.find('UserRankingCard').nodes[3].key).toBe('user1');
    });

    it('adds the class is-active to "Sort by ranking" if sortBy is "Sort by ranking"', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.find('.ranking').nodes[0].props.className).toBe('dropdown-item ranking is-active');
        expect(wrapper.find('.comments').nodes[0].props.className).toBe('dropdown-item comments ');
        expect(wrapper.find('.comment-votes').nodes[0].props.className).toBe('dropdown-item comment-votes ');
        expect(wrapper.find('.article-votes').nodes[0].props.className).toBe('dropdown-item article-votes ');
    });

    it('adds the class is-active to "Sort by comments" if sortBy is "Sort by comments"', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        wrapper.setState({ sortBy: 'Sort by comments' });
        expect(wrapper.find('.ranking').nodes[0].props.className).toBe('dropdown-item ranking ');
        expect(wrapper.find('.comments').nodes[0].props.className).toBe('dropdown-item comments is-active');
        expect(wrapper.find('.comment-votes').nodes[0].props.className).toBe('dropdown-item comment-votes ');
        expect(wrapper.find('.article-votes').nodes[0].props.className).toBe('dropdown-item article-votes ');
    });

    it('adds the class is-active to "Sort by comment votes" if sortBy is "Sort by comment votes"', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        wrapper.setState({ sortBy: 'Sort by comment votes' });
        expect(wrapper.find('.ranking').nodes[0].props.className).toBe('dropdown-item ranking ');
        expect(wrapper.find('.comments').nodes[0].props.className).toBe('dropdown-item comments ');
        expect(wrapper.find('.comment-votes').nodes[0].props.className).toBe('dropdown-item comment-votes is-active');
        expect(wrapper.find('.article-votes').nodes[0].props.className).toBe('dropdown-item article-votes ');
    });

    it('adds the class is-active to "Sort by article votes" if sortBy is "Sort by article votes"', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        wrapper.setState({ sortBy: 'Sort by article votes' });
        expect(wrapper.find('.ranking').nodes[0].props.className).toBe('dropdown-item ranking ');
        expect(wrapper.find('.comments').nodes[0].props.className).toBe('dropdown-item comments ');
        expect(wrapper.find('.comment-votes').nodes[0].props.className).toBe('dropdown-item comment-votes ');
        expect(wrapper.find('.article-votes').nodes[0].props.className).toBe('dropdown-item article-votes is-active');
    });

    it('changes the dropdown to is-active if this.state.showDropdown', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.find('.dropdown').nodes[0].props.className).toBe('dropdown ');
        wrapper.setState({ showDropdown: true });
        expect(wrapper.find('.dropdown').nodes[0].props.className).toBe('dropdown is-active');
    });

    it('triggers the sortUserList function, which changes the state, if the Sort by ranking is clicked', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        wrapper.setState({ sortBy: 'Sort by article votes' });
        expect(wrapper.state('sortBy')).toBe('Sort by article votes');

        wrapper.find('.ranking').simulate('click', { preventDefault() { }, target: { innerText: 'Sort by ranking' } });
        expect(wrapper.state('sortBy')).toBe('Sort by ranking');
    });

    it('triggers the sortUserList function, which changes the state, if the Sort by comments is clicked', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.state('sortBy')).toBe('Sort by ranking');

        wrapper.find('.comments').simulate('click', { preventDefault() { }, target: { innerText: 'Sort by comments' } });
        expect(wrapper.state('sortBy')).toBe('Sort by comments');
    });

    it('triggers the sortUserList function, which changes the state, if the Sort by comment votes is clicked', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.state('sortBy')).toBe('Sort by ranking');

        wrapper.find('.comment-votes').simulate('click', { preventDefault() { }, target: { innerText: 'Sort by comment votes' } });
        expect(wrapper.state('sortBy')).toBe('Sort by comment votes');
    });

    it('triggers the sortUserList function, which changes the state, if the Sort by article votes is clicked', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.state('sortBy')).toBe('Sort by ranking');

        wrapper.find('.article-votes').simulate('click', { preventDefault() { }, target: { innerText: 'Sort by article votes' } });
        expect(wrapper.state('sortBy')).toBe('Sort by article votes');
    });

    it('triggers the toggleDropdown function, which changes the state and changes the class of the dropdown, if the cancel option is clicked', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.state('showDropdown')).toBe(false);
        expect(wrapper.find('.dropdown').nodes[0].props.className).toBe('dropdown ');

        wrapper.find('.dropdown-button').simulate('click');
        expect(wrapper.state('showDropdown')).toBe(true);
        expect(wrapper.find('.dropdown').nodes[0].props.className).toBe('dropdown is-active');

        wrapper.find('.cancel').simulate('click');
        expect(wrapper.state('showDropdown')).toBe(false);
        expect(wrapper.find('.dropdown').nodes[0].props.className).toBe('dropdown ');
    });

    it('shows a loading icon if usersLoading is true', () => {
        const wrapperA = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapperA.find('.fa-spin').length).toBe(0);

        const wrapperB = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={true}
        />);
        expect(wrapperB.find('.fa-spin').length).toBe(1);
    });

    it('renders the UserRankingCard component as many times as there are users in the array', () => {
        const wrapper = shallow(<UserListCard
            users={users}
            fetchUsers={x => x}
            usersLoading={false}
        />);
        expect(wrapper.find('UserRankingCard').length).toBe(4);
    });
});