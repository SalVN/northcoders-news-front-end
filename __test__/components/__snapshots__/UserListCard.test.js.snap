// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListCard renders correctly 1`] = `
<div
  className="card-box"
>
  <div
    className="container-fluid"
  >
    <div
      className="card"
    >
      <header
        className="card-header"
      >
        <div
          className="card-header-style"
        >
          <div
            className="card-header-title"
          >
            <div
              className="dropdown "
            >
              <div
                className="dropdown-trigger"
              >
                <button
                  aria-controls="dropdown-menu"
                  aria-haspopup="true"
                  className="button is-small is-primary is-inverted is-outlined dropdown-button"
                  onClick={[Function]}
                >
                  <span
                    className="icon"
                  >
                    <i
                      aria-hidden="true"
                      className="fa fa-caret-down"
                    />
                  </span>
                </button>
              </div>
              <div
                className="dropdown-menu"
                id="dropdown-menu"
                role="menu"
              >
                <div
                  className="dropdown-content"
                >
                  <a
                    className="dropdown-item ranking is-active"
                    onClick={[Function]}
                  >
                    Sort by ranking
                  </a>
                  <a
                    className="dropdown-item comments "
                    onClick={[Function]}
                  >
                    Sort by comments
                  </a>
                  <a
                    className="dropdown-item comment-votes "
                    onClick={[Function]}
                  >
                    Sort by comment votes
                  </a>
                  <a
                    className="dropdown-item article-votes "
                    onClick={[Function]}
                  >
                    Sort by article votes
                  </a>
                  <a
                    className="dropdown-item cancel"
                    onClick={[Function]}
                  >
                    Close
                  </a>
                </div>
              </div>
            </div>
            <span>
              User Ranking
            </span>
          </div>
        </div>
      </header>
      <div
        className="card-content"
      >
        <div
          className="container-fluid mini-card"
        >
          <div
            className="columns is-mobile"
          >
            <div
              className="column is-1 is-hidden-tablet-only"
            >
              <p
                className="user-list-card-number"
              >
                <strong>
                  1
                </strong>
              </p>
            </div>
            <div
              className="column is-two-thirds"
            >
              <a
                href="/users/user4"
                onClick={[Function]}
              >
                <div>
                  <img
                    className="user-list-image"
                    src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
                  />
                  <span
                    className="user-list-username-text"
                  >
                    user4
                  </span>
                </div>
              </a>
            </div>
            <div
              className="column is-3"
            >
              <p
                className="user-list-card-ranking-number"
              >
                75
              </p>
            </div>
          </div>
        </div>
        <div
          className="container-fluid mini-card"
        >
          <div
            className="columns is-mobile"
          >
            <div
              className="column is-1 is-hidden-tablet-only"
            >
              <p
                className="user-list-card-number"
              >
                <strong>
                  2
                </strong>
              </p>
            </div>
            <div
              className="column is-two-thirds"
            >
              <a
                href="/users/user3"
                onClick={[Function]}
              >
                <div>
                  <img
                    className="user-list-image"
                    src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
                  />
                  <span
                    className="user-list-username-text"
                  >
                    user3
                  </span>
                </div>
              </a>
            </div>
            <div
              className="column is-3"
            >
              <p
                className="user-list-card-ranking-number"
              >
                69
              </p>
            </div>
          </div>
        </div>
        <div
          className="container-fluid mini-card"
        >
          <div
            className="columns is-mobile"
          >
            <div
              className="column is-1 is-hidden-tablet-only"
            >
              <p
                className="user-list-card-number"
              >
                <strong>
                  3
                </strong>
              </p>
            </div>
            <div
              className="column is-two-thirds"
            >
              <a
                href="/users/user2"
                onClick={[Function]}
              >
                <div>
                  <img
                    className="user-list-image"
                    src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
                  />
                  <span
                    className="user-list-username-text"
                  >
                    user2
                  </span>
                </div>
              </a>
            </div>
            <div
              className="column is-3"
            >
              <p
                className="user-list-card-ranking-number"
              >
                66
              </p>
            </div>
          </div>
        </div>
        <div
          className="container-fluid mini-card"
        >
          <div
            className="columns is-mobile"
          >
            <div
              className="column is-1 is-hidden-tablet-only"
            >
              <p
                className="user-list-card-number"
              >
                <strong>
                  4
                </strong>
              </p>
            </div>
            <div
              className="column is-two-thirds"
            >
              <a
                href="/users/user1"
                onClick={[Function]}
              >
                <div>
                  <img
                    className="user-list-image"
                    src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
                  />
                  <span
                    className="user-list-username-text"
                  >
                    user1
                  </span>
                </div>
              </a>
            </div>
            <div
              className="column is-3"
            >
              <p
                className="user-list-card-ranking-number"
              >
                50
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
